media-source-respec.html

```diff
@@ -663,7 +637,25 @@ <h3>Attributes</h3>
             <li>If {{MediaSource/[[live seekable range]]}} contains a range, then set {{MediaSource/[[live seekable
               range]]}} to be a new empty {{TimeRanges}} object.
           </li></ol>
-        <div><em>No parameters.</em></div><div><em>Return type: </em>{{undefined}}</div></dd><dt><dfn><code>isTypeSupported</code></dfn>, static</dt><dd>
+        <div><em>No parameters.</em></div><div><em>Return type: </em>{{undefined}}</div></dd>
+
+      <dt><dfn><code>getHandle</code></dfn></dt><dd>
+        <p>Creates a handle useful for attachment to an {{HTMLMediaElement}} via {{HTMLMediaElement/srcObject}}.</p>
+        <ol class="method-algorithm">
+          <li>If the implementation does not support creating a handle for this {{MediaSource}}, then throw a
+            {{NotSupportedError}} exception and abort these steps.
+            <p class="note">Implementations MAY choose to only allow handle creation for {{MediaSource}} objects in a
+              {{DedicatedWorkerGlobalScope}}, as a minimum requirement for enabling attachment of such a {{MediaSource}}
+              object to an {{HTMLMediaElement}}.</p>
+            <p class="issue">Consider requiring conforming implementations to let such handle be created on a {{Window}}
+              {{MediaSource}} object, too?</p>
+          </li>
+          <li>Create a new {{MediaSourceHandle}} object and associated resources.</li>
```

Comment:
Hmm. In the objectURL main-thread current way, you can obtain multiple object URLs for the same MediaSource. If you then attempt synchronous start of attachment to multiple media elements using either one of these multiple times, or the variety of these urls, at most one may succeed.

IOW, the current mechanism allows for multiple urls and simultaneous attempts to attach to the underlying mediasource, but doesn't let more than one of these succeed per underlying mediasource.

With auto-url-revocation upon successful attachment start of MSE object urls logic (as was added in Chrome), a single MSE objectURL cannot be successfully used in multiple attachments.

w.r.t. the new handle for attachment, I'd like to understand your concern better:
1) Would a single handle for a single MediaSource be allowed to attach more than once (assuming the underlying MediaSource returns to "closed" readyState between the attachments as per the detach logic)? Or should a single handle for a single MediaSource be allowed at most one successful attachment?

2) Would multiple distinct handles be retrievable for the same underlying MediaSource, and be usable for attachment so long as, at the attachment attempt time, the underlying MediaSource is "closed"?

(These questions are orthogonal to the further restriction that implementations are not yet required to support main thread MediaSource handle's attachment success ever, though they MAY in the current draft.)

Reply:
> IOW, the current mechanism allows for multiple urls and simultaneous attempts to attach to the underlying mediasource, but doesn't let more than one of these succeed per underlying mediasource.

I'm assuming the goal is not to support the concurrent connection of multiple media elements to a single `MediaSource`.  (Perhaps this could be achieved, given the largely one-directional flow of information from MediaSource to HTMLMediaElement, but the `MediaSource#readyState` model at least would need some adjustment.)

You draw an accurate distinction between multiple urls and my expectation of only a single handle.  The main-thread has the distinction of no concurrent operations and the ease of synchronously checking for an existing attachment.

> With auto-url-revocation upon successful attachment start of MSE object urls logic (as was added in Chrome), a single MSE objectURL cannot be successfully used in multiple attachments.

I guess this model would depend on a global mutex, even shared across processes.
I'd prefer to aim for something more deterministic that could be implemented with a message-passing design, if that would meet the needs.

> Would a single handle for a single MediaSource be allowed to attach more than once (assuming the underlying MediaSource returns to "closed" readyState between the attachments as per the detach logic)? Or should a single handle for a single MediaSource be allowed at most one successful attachment?

I'm anticipating that a single `MediaSourceHandle` would be allowed to be attached to at most one media element at any one time.

I haven't put too much thought into permitting detaching from one media element and reattaching to another, but I guess this could be supported, if desired.

> Would multiple distinct handles be retrievable for the same underlying MediaSource, and be usable for attachment so long as, at the attachment attempt time, the underlying MediaSource is "closed"?

I'm expecting not.  That's because of the difficultly in deterministically preventing "Attaching to a media element" when another attachment might be concurrently in progress.  Each such handle could be transferred to a different realm, having a different event loop, and so attachment attempts could occur concurrently.

Even if two handles were transferred to the same realm, the asynchronicity of 'Set the `readyState` to "open"' means that `srcObject` might be set on two media elements before `readyState` changes.  I dare say that attachment failure could be reported asynchronously, but this would not be deterministic in the case of realms having different event loops, and API that prevented such failure attempts or caught them synchronously (or even as a promise rejection, if necessary) would be preferable to async error events.

Reply:
Simultaneous, concurrent usage of the same underlying MediaSource with different media elements is certainly not the intent.

Given we have this opportunity to reduce developer confusion about this intent (versus the more general object URL creation/use idiom), I tend to agree that simplifying to only allow one handle per MediaSource would be a good idea.

However, I'm less clear about how to specify what happens once such handle is transferred from worker context to main context: should the mediaSource.getHandle() method fail (in worker) with some kind of exception?

Note, that the handle itself could know whether or not it is currently involved in an attachment to a media element, such that the MSE-specific spec text for attaching and detaching such a handle can precisely ensure that at most one element can currently have the handle as the "assigned media object" (or otherwise do the appropriate attachment failure steps on the attempts to attach an already attached handle).

This would mirror the expected behavior for eventual specification of support for main thread-owned MediaSource attachments to a media element via srcObject (not in scope of this PR work): such a main thread MediaSource could only ever be attached successfully if it isn't already a media provider object (or already being used in an attachment via the dereferenced MediaSource if attached by a main-thread-originated MediaSource objectURL).

@karlt If you agree with this approach, I'll land this PR and create a follow-up that employs this restriction (since my co-editor has already approved this PR). (Heads-up, @mwatson2 ). Also, thank you both for your assistance on this.

Reply:
> Given we have this opportunity to reduce developer confusion about this intent (versus the more general object URL creation/use idiom),

That's an excellent explanation.

> However, I'm less clear about how to specify what happens once such handle is transferred from worker context to main context: should the mediaSource.getHandle() method fail (in worker) with some kind of exception?

Yes, I would expect `mediaSource.getHandle()` to throw when called a second time, and continue to throw after the handle from the first call has been transferred.

When transferred from Worker to Window, the handle in worker context would become neutered, if that is the right term, so that it could not be transferred to multiple recipients.

> Note, that the handle itself could know whether or not it is currently involved in an attachment to a media element, such that the MSE-specific spec text for attaching and detaching such a handle can precisely ensure that at most one element can currently have the handle as the "assigned media object" (or otherwise do the appropriate attachment failure steps on the attempts to attach an already attached handle).

Yes, that should be very workable, and would be a simple and clear concept for developers to follow.

> @karlt If you agree with this approach, I'll land this PR and create a follow-up that employs this restriction (since my co-editor has already approved this PR).

Yes, please go ahead and land this.  Thank you for writing it up.

Reply:
@karlt thank you for your comments. I'll land this now.

Reply:
But first I'll rebase over the already merged prerequisite.